substrate/frame/utility/src/lib.rs

```diff
@@ -454,6 +465,71 @@ pub mod pallet {
 			let res = call.dispatch_bypass_filter(frame_system::RawOrigin::Root.into());
 			res.map(|_| ()).map_err(|e| e.error)
 		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight({
+			let main_dispatch_info = main.get_dispatch_info();
+			let fallback_dispatch_info = fallback.get_dispatch_info();
+			(
+				T::WeightInfo::if_else()
+					.saturating_add(main_dispatch_info.call_weight)
+					.saturating_add(fallback_dispatch_info.call_weight),
+				main_dispatch_info.class,
+			)
+		})]
+		pub fn if_else(
+			origin: OriginFor<T>,
+			main: Box<<T as Config>::RuntimeCall>,
+			fallback: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResultWithPostInfo {
+			// Do not allow the `None` origin.
+			if ensure_none(origin.clone()).is_ok() {
+				return Err(BadOrigin.into());
+			}
+
+			let is_root = ensure_root(origin.clone()).is_ok();
+
+			// Track the weights
+			let mut weight = Weight::zero();
+
+			let info = main.get_dispatch_info();
+
+			// Execute the main call first
+			let main_result = if is_root {
+				main.dispatch_bypass_filter(origin.clone())
+			} else {
+				main.dispatch(origin.clone())
+			};
+
+			// Add weight of the main call
+			weight = weight.saturating_add(extract_actual_weight(&main_result, &info));
+
+			if let Err(_main_call_error) = main_result {
+				// If the main call failed, execute the fallback call
+				let fallback_info = fallback.get_dispatch_info();
+
+				let fallback_result = if is_root {
+					fallback.dispatch_bypass_filter(origin.clone())
+				} else {
+					fallback.dispatch(origin.clone())
+				};
+
+				// Add weight of the fallback call
+				weight =
+					weight.saturating_add(extract_actual_weight(&fallback_result, &fallback_info));
+
+				if let Err(_fallback_error) = fallback_result {
+					// Both calls have faild.
+					return Err(Error::<T>::InvalidCalls.into())
```

Comment:
It would be better to have 3 events: main success, main error and fail success, main error and fail error.

Reply:
```rust
/// if_else main call was dispatched.
IfElseMainSuccess,
/// if_else fallback call was dispatched.
IfElseFallbackSuccess { main_error: DispatchError }, 
/// if_else both calls failed.
IfElseFailure { main_error: DispatchError, fallback_error: DispatchError },
```

Reply:
Yes great

Reply:
`IfElseFallbackFailure` or `IfElseBothFailure` for the last variant?